### PR TITLE
Enable extending extra helm parameters

### LINKED ISSUE #40: Support custom theme
## Context
In order to provide a custom theme [according to the codecentric documentation](https://github.com/codecentric/helm-charts/blob/master/charts/keycloak/README.md#providing-a-custom-theme) we have to add extraInitContainers, extraVolumeMounts, and extraVolumes. 
In the component-keycloak that should be supported easily. At the moment we to copy the whole extraVolumes String ([see component-keycloak/defaults.yml](https://github.com/projectsyn/component-keycloak/blob/532880ab59f6966d0ebe1253d9ee13eb5a83f6d5/class/defaults.yml#L161-L179)) to our cluster definition file to add an additional volume.

For https://github.com/vshn/keycloak-theme/ we have to add the values below, but we also have to copy everything from [defaults.yml](https://github.com/projectsyn/component-keycloak/blob/master/class/defaults.yml#L161-L179), which leads to duplication. 
```yaml
      extraVolumes: |
        - name: db-certs
          secret:
            secretName: keycloak-postgresql-tls
            defaultMode: 0400
            items:
              - key: tls.crt
                path: tls.crt
        - name: keycloak-tls
          secret:
            secretName: keycloak-tls
            defaultMode: 420
        - name: theme
          emptyDir: {}
      extraVolumeMounts: |
        - name: db-certs
          readOnly: true
          mountPath: /opt/jboss/certs
        - name: keycloak-tls
          readOnly: true
          mountPath: /etc/x509/https
        - name: theme
          readOnly: true
          mountPath: /opt/jboss/keycloak/themes/vshn
      extraInitContainers: |
        - name: theme-provider
          image: quay.io/vshn/keycloak-theme:v1.0.0
          imagePullPolicy: IfNotPresent
          command:
            - sh
          args:
            - -c
            - |
              echo "Copying theme..."
              cp -R /theme/* /vshn-theme
          volumeMounts:
            - name: theme
              mountPath: /vshn-theme
```

## Acceptance criteria

* We only should only need to define the image name and mount points as parameters.
* Add a small how-to page describing how to develop a custom theme.

## Alternatives
Copy paste all extraVolume definitions from component-keycloak works but.

### LINKED ISSUE #44: Support Custom Environment Varibles
## Context
In order to provide a custom theme in the welcome screen, we need to set `KEYCLOAK_WELCOME_THEME` or `KEYCLOAK_DEFAULT_THEME` (see https://hub.docker.com/r/jboss/keycloak/)

In the component-keycloak that should be supported easily.

At the moment it looks like simiilar to [issue #40](https://github.com/projectsyn/component-keycloak/issues/40) and we would have to copy the whole extraEnv String (see [component-keycloak/defaults.yml](https://github.com/projectsyn/component-keycloak/blob/master/class/defaults.yml#L135-L155)) to our cluster definition file to add an additional volume.

### PR DESCRIPTION
**Pull request by glrf**
## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Link this PR to related issues.

closes #40 
closes #44

### CODE REVIEW

File: tests/builtin.yml [hunk 1]
```diff
@@ -1 +1,20 @@
 ---
+parameters:
+  keycloak:
+    extraEnv:
+      "FOO":
+        value: "bar"
+    extraInitContainers:
+      "theme-provider":
```
Comment (ccremer) [inline tests/builtin.yml:8]: Those quotes aren't required AFAIK
```suggestion
      FOO:
        value: bar
    extraInitContainers:
      theme-provider:
```

Comment (glrf) [reply tests/builtin.yml:8]: Yes probably not. I'm generally use quotes a bit too liberally :)

But in general I think this parameter structure is a lot better then more strings, and the postprocessing step is not too bad.

File: postprocess/extra-env.jsonnet [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+{
+  statefulset: statefulset {
+    spec+: {
+      template+: {
+        spec+: {
+          containers: [
+            if c.name == 'keycloak' then
+              c {
+                env: insertMap(
+                  if 'env' in super then super.env else [],
```
Comment (glrf) [inline postprocess/extra-env.jsonnet:31]: If someone has a better solution to check if the field exists in `super`, I'm all ears.

Comment (simu) [reply postprocess/extra-env.jsonnet:31]: Generally, `std.objectHas` should work with `super`

```suggestion
                  if std.objectHas(super, 'env') then super.env else [],
```

Or you could use the commodore.libjsonnet wrapper:

```suggestion
                  com.getValueOrDefault(super, 'env', []),
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 3]
```diff
@@ -352,6 +352,92 @@ parameters:
...
+        value: bar
+----
+
+== `extraVolumes`
+
+[horizontal]
+type:: dict
+default:: {}
+
+Extra volumes added to the keycloak statefulset.
```
Comment (simu) [inline docs/modules/ROOT/pages/references/parameters.adoc:380]:
```suggestion
Extra volumes added to the Keycloak statefulset.
Keys in the dict are used as value for field `name` in the resulting volume configuration
Values must be valid Kubernetes volume configurations.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 4]
```diff
@@ -352,6 +352,92 @@ parameters:
...
+        emptyDir: {}
+----
+
+== `extraVolumeMounts`
+
+[horizontal]
+type:: dict
+default:: {}
+
+Extra volumemounts added to the keycloak container.
```
Comment (simu) [inline docs/modules/ROOT/pages/references/parameters.adoc:399]:
```suggestion
Extra volume mounts added to the Keycloak container.
Keys in the dict are used as value for field `name` in the resulting volume mount configuration
Values must be valid Kubernetes volume mount configurations.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 5]
```diff
@@ -352,6 +352,92 @@ parameters:
...
+        mountPath: /opt/jboss/keycloak/themes/vshn
+----
+
+== `extraInitContainers`
+
+[horizontal]
+type:: dict
+default:: {}
+
+Extra init containers added to the keycloak statefulset.
```
Comment (simu) [inline docs/modules/ROOT/pages/references/parameters.adoc:418]:
```suggestion
Extra init containers added to the keycloak statefulset.
Keys in the dict are used as value for field `name` in the resulting container configuration
Values must be valid Kubernetes container configurations.
```

**Review by simu — APPROVED** [review]
LGTM overall. Documentation for the new parameters is very sparse, I've added a suggestion to make the docs more clearer in regard to expected values and rendering performed by the component.

File: postprocess/extra-env.jsonnet (hunk 2 above)
Comment (glrf) [reply postprocess/extra-env.jsonnet:31]: Sadly this fails for me with 
`RuntimeError: STATIC ERROR: /keycloak/postprocess/extra-env.jsonnet:31:36-41: expected . or [ after super.`

That's why i switched to using `in`

Comment (simu) [reply postprocess/extra-env.jsonnet:31]: Theoretically `super` and `c` should be equivalent in this context, cf. https://jsonnet.org/ref/language.html#inheritance

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 6]
```diff
@@ -352,6 +352,100 @@ parameters:
...
 
+== `extraEnv`
+
+[horizontal]
+type:: dict
+default:: {}
+
+Extra environment variables added to the Keycloak statefulset.
+Keys in the dict are used as value for field `name` in the resulting environment variable configuration
+Values must be valid Kubernetes environment variable configurations.
```
Comment (ccremer) [inline docs/modules/ROOT/pages/references/parameters.adoc:363]:
```suggestion
Extra environment variables added to the Keycloak StatefulSet.
Keys in the dict are used as value for field `name` in the resulting environment variable configuration.
Values must be valid Kubernetes environment variable configurations.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 7]
```diff
@@ -352,6 +352,100 @@ parameters:
...
+
+== `extraVolumes`
+
+[horizontal]
+type:: dict
+default:: {}
+
+Extra volumes added to the Keycloak statefulset.
+Keys in the dict are used as value for field `name` in the resulting volume configuration
+Values must be valid Kubernetes volume configurations.
```
Comment (ccremer) [inline docs/modules/ROOT/pages/references/parameters.adoc:384]:
```suggestion
Extra volumes added to the Keycloak StatefulSet.
Keys in the dict are used as value for field `name` in the resulting volume configuration.
Values must be valid Kubernetes volume configurations.
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 8]
```diff
@@ -352,6 +352,100 @@ parameters:
...
+        command:
+          - sh
+        args:
+          - -c
+          - |
+            echo "Copying theme..."
+            cp -R /theme/* /vshn-theme
+        volumeMounts:
+          - name: theme
+            mountPath: /vshn-theme
```
Comment (ccremer) [inline docs/modules/ROOT/pages/references/parameters.adoc:446]: I would avoid mentioning VSHN
```suggestion
        image: company/keycloak-theme:v1.0.0
        imagePullPolicy: IfNotPresent
        command:
          - sh
        args:
          - -c
          - |
            echo "Copying theme..."
            cp -R /theme/* /company-theme
        volumeMounts:
          - name: theme
            mountPath: /company-theme
```

File: postprocess/extra-env.jsonnet [hunk 9]
```diff
@@ -0,0 +1,55 @@
+/**
+ * Adjust StatefuleSet generated by helm template
```
Comment (ccremer) [inline postprocess/extra-env.jsonnet:2]:
```suggestion
 * Adjust StatefulSet generated by helm template
```

File: tests/builtin.yml [hunk 10]
```diff
@@ -1 +1,29 @@
...
+        command:
+          - sh
+        args:
+          - -c
+          - |
+            echo "Copying theme..."
+            cp -R /theme/* /vshn-theme
+        volumeMounts:
+          - name: theme
+            mountPath: /vshn-theme
```
Comment (ccremer) [inline tests/builtin.yml:22]:
```suggestion
      theme-provider:
        image: company/keycloak-theme:v1.0.0
        imagePullPolicy: IfNotPresent
        command:
          - sh
        args:
          - -c
          - |
            echo "Copying theme..."
            cp -R /theme/* /company-theme
        volumeMounts:
          - name: theme
            mountPath: /company-theme
```

**Review by ccremer — CHANGES_REQUESTED** [review]

**Comment by glrf** [comment]
Added how-to on custom themes in #69

**Review by ccremer — APPROVED** [review]
LGTM. Not sure what you want to do with the open conversations though

File: postprocess/extra-env.jsonnet (hunk 2 above)
Comment (glrf) [reply postprocess/extra-env.jsonnet:31]: Are you fine with leaving this as is? As far as I can tell the current solution is correct and the other option of using `com.getValueOrDefault` is IMO not more readable, especially for the other cases where I cannot simply use `c`

Comment (simu) [reply postprocess/extra-env.jsonnet:31]: I'm fine to leave it as it is

**Pull request merged**